File: CS/WindowsApplication1/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraTab;
using DevExpress.XtraTab.ViewInfo;

namespace WindowsApplication1
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        XtraTabPage dragPage;
        DragWindow dragWindow;

        private void xtraTabControl1_MouseDown(object sender, MouseEventArgs e)
        {
            dragPage = (sender as XtraTabControl).CalcHitInfo(e.Location).Page;
        }

        private void xtraTabControl1_MouseMove(object sender, MouseEventArgs e)
        {
            if (dragPage != null & dragWindow == null)
                CreateDragWindow();
            if (dragWindow != null)
                dragWindow.Location = Control.MousePosition;
        }

        private void xtraTabControl1_MouseUp(object sender, MouseEventArgs e)
        {
            XtraTabHitInfo hitInfo = (sender as XtraTabControl).CalcHitInfo(e.Location);
            if (hitInfo.HitTest == XtraTabHitTest.PageHeader)
            {
                if (dragWindow != null)
                {
                    dragPage.Controls.Add(dragWindow.ClonedPage.Controls[0]);
                    dragWindow.Close();
                }
            }
            else
            {
                FinalizeDragWindow();
            }
            dragPage = null;
            dragWindow = null;
        }
        private void CreateDragWindow()
        {
            if (dragWindow == null)
                dragWindow = new DragWindow(dragPage);
        }

        private void FinalizeDragWindow()
        {
            if (dragWindow == null) return;
            xtraTabControl1.TabPages.Remove(dragPage);
            dragWindow.FormBorderStyle = FormBorderStyle;
            dragWindow.Opacity = 100;
            dragWindow.WindowState = FormWindowState.Maximized;
        }
    }
}

File: VB/WindowsApplication1/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraTab
Imports DevExpress.XtraTab.ViewInfo

Namespace WindowsApplication1
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private dragPage As XtraTabPage
		Private dragWindow As DragWindow

		Private Sub xtraTabControl1_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs) Handles xtraTabControl1.MouseDown
			dragPage = (TryCast(sender, XtraTabControl)).CalcHitInfo(e.Location).Page
		End Sub

		Private Sub xtraTabControl1_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs) Handles xtraTabControl1.MouseMove
			If dragPage IsNot Nothing And dragWindow Is Nothing Then
				CreateDragWindow()
			End If
			If dragWindow IsNot Nothing Then
				dragWindow.Location = Control.MousePosition
			End If
		End Sub

		Private Sub xtraTabControl1_MouseUp(ByVal sender As Object, ByVal e As MouseEventArgs) Handles xtraTabControl1.MouseUp
			Dim hitInfo As XtraTabHitInfo = (TryCast(sender, XtraTabControl)).CalcHitInfo(e.Location)
			If hitInfo.HitTest = XtraTabHitTest.PageHeader Then
				If dragWindow IsNot Nothing Then
					dragPage.Controls.Add(dragWindow.ClonedPage.Controls(0))
					dragWindow.Close()
				End If
			Else
				FinalizeDragWindow()
			End If
			dragPage = Nothing
			dragWindow = Nothing
		End Sub
		Private Sub CreateDragWindow()
			If dragWindow Is Nothing Then
				dragWindow = New DragWindow(dragPage)
			End If
		End Sub

		Private Sub FinalizeDragWindow()
			If dragWindow Is Nothing Then
				Return
			End If
			xtraTabControl1.TabPages.Remove(dragPage)
			dragWindow.FormBorderStyle = FormBorderStyle
			dragWindow.Opacity = 100
			dragWindow.WindowState = FormWindowState.Maximized
		End Sub
	End Class
End Namespace